Replace fallback pipeline with joiners like concat, fill, interleave

karlhigley opened the pull request.

karlhigley commented on src/poprox_recommender/default.py:
```diff
@@ -83,27 +83,22 @@ def personalized_pipeline(num_slots: int, algo_params: dict[str, Any] | None = N
+    # Score and rank articles with diversification/calibration reranking
     pipeline.add(article_scorer, inputs=["candidate", "profile"], output="candidate")
-    pipeline.add(topk_ranker, inputs=["candidate", "profile"], output="ranked")
     pipeline.add(ranker, inputs=["candidate", "profile"], output="reranked")
```
What's missing here is a mechanism for making the branch of the pipeline above return an empty list of articles when there are no clicks. One way to do that:
* Make the user embedding identically zero when there are no clicks
* Then the articles should all get zero scores
* Add a filter that removes articles with zero scores from the list

karlhigley replied:
Thoughts on this @mdekstrand?

mdekstrand replied:
At the risk of a few more `if` statements (that the typechecker can hopefully help us with), I think I would be more inclined to use `None`, and then scores be NA (Pandas handles this tolerably well, NumPy is ok as well if we just use NaN as a sentinel value, which is what Pandas does under the hood at least until it's fully Arrow-based). The advantages to this:

* fast-track null compute instead of computing a lot of zeros expensively
* no data clearly distinguished from other sources of 0
* avoid risks of "testing for exactly zero" (although filtering for `>= 1.0e-6` would fix this)

Biggest downside I see: more logic to test for sentinels.

The NaN/NA design is what LensKit uses — scorers are supposed to return NA for anything they can't score, and the fallback scorer just asks each scorer in turn to score evertything left NA by the previous ones. I would be even happier with this solution if Pandas had a proper NA like R instead of repurposing NaN, but it is the Pandas way (Pandas does have masked array types, but they aren't as widely used; raw NumPy does not have masked arrays; Arrow does).

karlhigley replied:
Yeah, those are good points. I'll head that direction!

karlhigley commented on src/poprox_recommender/default.py:
```diff
@@ -83,27 +83,22 @@ def personalized_pipeline(num_slots: int, algo_params: dict[str, Any] | None = N
     pipeline.add(topic_filter, inputs=["candidate", "profile"], output="topical")
-    pipeline.add(sampler, inputs=["topical", "candidate"], output="ranked")
+    pipeline.add(sampler, inputs=["topical", "candidate"], output="sampled")
+    pipeline.add(fill, inputs=["reranked", "sampled"], output="recs")
```
Once the pipeline itself supports fallbacks, we might consider whether we want to use `Fill` or a fallback here

mdekstrand commented on src/poprox_recommender/components/joiners/concat.py:
```diff
@@ -0,0 +1,9 @@
+        reverse_articles = (candidates1.articles + candidates2.articles)[::-1]
+        articles = {article.article_id: article for article in reverse_articles}
+
+        return ArticleSet(articles=list(articles.values())[::-1])
```
I think I see how this code works, but I don't think it's very obvious to someone who isn't familiar with ins and outs of dictionary ordering ordering and reversing. Maybe some comments describing why we build the dictionary in reverse order, and the reverse it again, would help?

mdekstrand replied:
for consistency with the above change, this could also be `list(reversed(articles.values()))`.

mdekstrand commented on src/poprox_recommender/components/joiners/concat.py:
```diff
@@ -0,0 +1,9 @@
+
+class Concatenate:
+    def __call__(self, candidates1: ArticleSet, candidates2: ArticleSet) -> ArticleSet:
+        reverse_articles = (candidates1.articles + candidates2.articles)[::-1]
+        articles = {article.article_id: article for article in reverse_articles}
```
Might this be more readable (less dependent on list indexing magic) if it is:

```python
articles = {article.article_id: article for article in reversed(candidates1.articles + candidates2.articles)}
```

This would also have advantage of reducing one list allocation, but that isn't a sufficient reason to do it.

mdekstrand replied:
Mostly thinking that explicit use of `reversed` may be easier to understand at first look for less-experienced Python programmers — it does exactly what it says.

karlhigley replied:
It's interesting: I typically don't use nearly as much indexing magic on my own as the students seem to reach for

mdekstrand commented on src/poprox_recommender/components/joiners/fill.py:
```diff
@@ -0,0 +1,15 @@
+        articles = candidates1.articles[: self.num_slots]
+
+        # Fill out remaining slots from candidates2
+        articles += candidates2.articles[: self.num_slots - len(articles)]
```
what happens when the index is negative here? (I do not remember that nuance of Python indexing.)

mdekstrand replied:
For consideration — would it be easier with itertools? `list(islice(chain(candidates1.articles, candidates2.articles), self.num_slots))` — then all the edge cases around the lengths of the two article lists are handled by the library.

I don't have a strong opinion on relative readability, but making the code “obviously correct” in the face of edge cases (the key one being when `candidates1` has more than `num_slots` articles) may add complexity.

karlhigley replied:
I'm generally good with improving the edge case handling of this code but it's not obvious to me what the itertools snippet is actually doing without the itertools docs in front of me 😅

mdekstrand replied:
At the minor cost of more list allocations, we could do list combination:

```python
articles = candidates1.articles + candidates2.articles
articles = articles[:self.num_slots]
```

mdekstrand commented on src/poprox_recommender/components/joiners/interleave.py:
```diff
@@ -0,0 +1,10 @@
+    def __call__(self, candidates1: ArticleSet, candidates2: ArticleSet) -> ArticleSet:
+        articles = [None] * (len(candidates1.articles) + len(candidates2.articles))
+        articles[::2] = candidates1.articles
+        articles[1::2] = candidates2.articles
```
If you're interested in going down the itertools route, itertools doesn't have interleave but more-itertools does.

karlhigley replied:
Of the indexing magic contained here, this is the one I think probably makes the most sense

mdekstrand reviewed (commented):
Mostly readability + 1 edge-case correctness comment.

karlhigley commented:
Took different approaches to improving readability in the three cases:
* For `Concatenate`, I can't think of a better way to do what I'm aiming at so I added a lengthy comment explaining what's happening
* For `Fill`, I adopted your list combination and then slicing approach
* For `Interleave`, I rewrote it with `itertools.zip_longest` which seems pretty intuitive if you know `zip`
